### ISSUE TITLE
Missing "Task" case

### ISSUE THREAD

**Issue by SeRG1k17** [Moya/Moya#1647]
In one of my projects, I needed to use the parameters in the query string, and also use the body.  Due to the fact that the property path can not be represented as:
```swift
public var path: String {
        switch self {
case .myCase(let name): return "api/endpoint?username=\(name)"
}
```
Here's what my case should look like:
```swift
case requestJSONEncodable(Encodable, parameters: [String: Any], encoding: ParameterEncoding)
```

**Comment by sunshinejr** [comment]
I think that this proposal makes sense. We usually name the cases using `composable` keyword, so in this case, it could be something similar to `requestCompositeJSONEncodable(encodable: Encodable, urlParameters:  [String: Any])`. The implementation could also be similar to [`requestCompositeData`](https://github.com/Moya/Moya/blob/master/Sources/Moya/Task.swift#L22).

I'd love to hear other @Moya/contributors what do they think about it :)

**Comment by ashfurrow** [comment]
Wait, I'm not quite clear on the problem. Both your request body and query string need the same parameters? Could the [`requestClosure`](https://github.com/Moya/Moya/blob/master/docs/Endpoints.md#request-mapping) be used for this? Or could the logic be abstracted into shared code?

**Comment by SD10** [comment]
I also want to make the argument that Moya should allow users to handle the mapping of an `Endpoint` to a `URLRequest` from within the `Endpoint` object itself

The mapping is currently defined as a `public` method:
https://github.com/Moya/Moya/blob/4555c7956e5acfb81bf759476c905f0866e5df4c/Sources/Moya/Endpoint.swift#L78

You can use the `requestClosure` of `MoyaProvider` to handle this mapping, however, this closure receives an `Endpoint` so the underlying `TargetType` has already been abstracted away. This makes it pretty provide any additional context that is available in the `TargetType` and not `Endpoint`.

You would need to:
1. subclass `Endpoint` to provide the additional context you want from `TargetType`
2. then provide your own `endpointClosure` to map to your new subclass
3. then provide your own `requestClosure` to handle mapping to `Result<URLRequest, MoyaError>`. 

IMO the burden can be reduced by:
1. Changing `Endpoint.urlRequest()` from `public` to `open`
   - This eliminates the need for `step 3` above.
2. Adding a `urlRequestClosure` argument to the `Endpoint` initializer.
   - This allows the user to control mapping to a `URLRequest` directly inside of the `endpointClosure` (where they still have access to the underlying `TargetType`). No need for a `requestResult` closure and no need to subclass `Endpoint`.

Sorry to drop this bomb 😢

**Comment by ashfurrow** [comment]
@SD10 There is no need to apologize! We can work through these things together as they come up – and things like this _do_ come up, for every project.

**Comment by SeRG1k17** [comment]
@SD10, @ashfurrow @sunshinejr Off-topic. I found another improvement. For example: we need a different timout for each request. Sure we need our manager. However, a typical timer change is described in #743. And in the last comment the author put forward a hypothesis. On adding one more case to the enumeration.

I believe we should reconsider some of the typical concepts that most developers face. After all, using closures is not always elegant.

**Comment by SeRG1k17** [comment]
@ashfurrow @SD10 @sunshinejr  I need case for Decodable body and parameters as queryString.
Because "api/endpoint?username=\(name)" in path property not working due to encoding. And even if it worked, the solution is very bad, i think.
```swift
    case requestJSONEncodable(Encodable)
    case requestParameters(parameters: [String: Any], encoding: ParameterEncoding)
    case requestCompositeData(bodyData: Data, urlParameters: [String: Any])
    case requestCompositeParameters(bodyParameters: [String: Any], bodyEncoding: ParameterEncoding, urlParameters: [String: Any])
```
None of them allow us to use Encodable + [String: Any] as url parameters
```swift
Sure i can use:
    public var task: Task {
        switch self {
        case let .setGlucoseThresholds(_, glucoseThresholds):
            let encoded = try! JSONEncoder().encode(glucoseThresholds)
            return .requestCompositeData(bodyData: encoded, urlParameters: parameters)
        }
    }
```
But I do not think this is a good decision. Because, this is not the place where we need to encode Encodable to data.

Update my proposal case:
case requestJSONEncodable(Encodable, urlParameters: [String: Any])

**Comment by SeRG1k17** [comment]
@SD10 I want to not replace, but add a case.

**Comment by sunshinejr** [comment]
So did we come back to the [solution I gave at the beginning](https://github.com/Moya/Moya/issues/1647#issuecomment-383342141)? 😄 I also do agree that we shouldn't really replace the case with additional parameter 👍

**Comment by SeRG1k17** [comment]
@sunshinejr Look above, friend 😃

**Comment by sunshinejr** [comment]
@SeRG1k17 Yeah, yeah, I just said that we all agree that replacing the current task is a no-no, and asked if you guys like the solution I gave at the beginning 😄

**Comment by SeRG1k17** [comment]
@sunshinejr Oh sorry. I did not notice. Yes, in fact, this is the same thing that you wrote above.

**Comment by SD10** [comment]
@SeRG1k17 I'm sympathetic to not being able to handle the encoding error. That is in fact, a real problem 🤔 

Can you add the parameters in the `baseURL` directly by making it dynamic and then ignore the path component for this case?

I just think that if we're going to add another case for this, we need to decide where we either draw the line or need to rethink things 😭

**Comment by SeRG1k17** [comment]
@SD10 I can do that. But this seems like a "crutch".
```swift
    public var baseURL: URL {
        
        var url = "https://example.com"
        
        switch self {
        case .custom(let value): url += "?value=value"
        default: break
        }
        return URL(string: url)!
    }
```

**Comment by ashfurrow** [comment]
Could this be something that Parameter Encoding solves? We have docs here: https://github.com/Moya/Moya/blob/master/docs/Examples/ParameterEncoding.md Basically you can provide a custom bit of code that translates the parameters of the endpoint into the request. Again, a bit of a clutch, but as pointed out this is an uncommon scenario. 

Moya should make common things easy and uncommon things possible, and a "clutch" in this case might be worth it for the sake of a smaller API surface area.

**Comment by pedrovereza** [comment]
On a side note, can we have this in the Vision document?

> Moya should make common things easy and uncommon things possible

**Comment by ashfurrow** [comment]
Good idea: https://github.com/Moya/Moya/pull/1659

**Comment by SD10** [comment]
This is a really long issue but for the sake of record keeping I want to make a few notes:
- You _cannot_ do this through the `requestClosure` because the `TargetType` has been abstracted
- You _can_ do this through the use of a `PluginType` because you still have access to `TargetType`

@SeRG1k17 Addresses the issue of wanting to use `Encodable` to encode parameters and there is currently not a good way to handle such an error. 

However, if we made the methods of the `PluginType` protocol throwing, then we could bubble up this error to the user when applying plugins.
